### DISCUSSION TITLE
advice: multiple source files for adaptive video quality

### DISCUSSION THREAD

**Discussion by sebalis** [videojs/video.js#7924] [Q&A] [answered]
Hi,

I am developing a small web app that will feature several simultaneously running video players. Each player will be served with a static video file. I cannot predict my users’ screen sizes and as the number of players might change in runtime their sizes might change too. I would therefore love to be able to be able to supply these video files in several qualitites (such as 480p, 720p, 1080p), and I hope that there would be some way of making the players figure out which quality to use, and perhaps make them “reconsider” this while the video is playing as the players’ sizes change. What I am not looking for is support for a playlist or a manual switcher.

I have good general experience with programming and web apps but not with web video players. I’m hoping that video.js as a well-established tool would have some feature, built-in or via a plugin, to achieve this. As a programming platform I am thinking of using React and Typescript. I have done some initial searching but am not sure if any of my findings are suitable. For example, this should not require specific streaming features on the web server because I don’t expect to be able to change the server configuration (and assume I would need to e.g. for DASH).

Could you give any advice on any of these points? Many thanks.

**Comment by gkatsev** [comment, marked as answer]
Your best bet is to make sure that your content is HLS (or DASH) for the various qualities. If they're regular mp4s, switching between them can't be seamless.
Video.js by default plays back HLS and will limit the quality to the player size.

**Comment by sebalis** [comment]
Thank you very much for responding, @gkatsev. Am I right in thinking that if I have a “dumb” server (i.e. configurable via .htaccess at most) on someone’s simplistic webspace, there is no way for me to make sure that my content is HLS or DASH? With some initial search results I’m not entirely sure. [This Stackoverflow page](https://stackoverflow.com/questions/37110482/how-to-serve-dash-video-mpeg-dash-and-hls-over-a-web-sever) suggests I need to set up a lot on the server, which I cannot do, while [this MDN page](https://developer.mozilla.org/en-US/docs/Web/Guide/Audio_and_video_delivery/Setting_up_adaptive_streaming_media_sources) suggests I just have to produce an MPD file (and I have seen an ffmpeg command line for that already) and then point the video player to that in the list of alternative sources.

**Reply by gkatsev** [reply]
Yeah, while you can serve HLS content from apache with .htaccess, but you'll need something to generate that HLS first and then upload it to your server.

**Reply by sebalis** [reply]
Ah, I now see that MPD is for DASH and HLS uses a different “playlist” format, which would be okay – but also that I would have to split my videos into fragments for either of these. I will probably not have the time to do that for my videos. Thanks for the answers and for pointing me to the detailed questions I had to research.

**Reply by gkatsev** [reply]
You definitely don't need to choose both. I would probably choose HLS only to choose just one because on iPhones you'll only get HLS. Also, HLS is a bit simpler.
It is possible to segment video so that they work for both DASH and HLS, but it's a bit trickier.

**Reply by sebalis** [reply]
That’s good to know, but in my case the fragmentation is probably too much effort even it is only for HLS. I would have too many videos to process and too little time. Perhaps this an imrprovement I should consider after the initial release though. Thanks again.

**Reply by gkatsev** [reply]
You're welcome!
